flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/OutputColsHelper.java

```diff
@@ -0,0 +1,200 @@
+	 */
+	private int[] outputColsPosInResult;
+
+	public OutputColsHelper(TableSchema inputSchema, String outputColName, TypeInformation outputColType) {
+		this(inputSchema, outputColName, outputColType, inputSchema.getFieldNames());
+	}
+
+	public OutputColsHelper(TableSchema inputSchema, String outputColName, TypeInformation outputColType,
+							String[] reservedColNames) {
+		this(inputSchema, new String[]{outputColName}, new TypeInformation[]{outputColType}, reservedColNames);
+	}
+
+	public OutputColsHelper(TableSchema inputSchema, String[] outputColNames, TypeInformation[] outputColTypes) {
+		this(inputSchema, outputColNames, outputColTypes, inputSchema.getFieldNames());
+	}
+
+	/**
+	 * The constructor.
+	 *
+	 * @param inputSchema      Schema of input data being predicted or transformed.
+	 * @param outputColNames   Output column names of the prediction/transformation operator.
+	 * @param outputColTypes   Output column types of the prediction/transformation operator.
+	 * @param reservedColNames Reserved column names, which is a subset of input data's column names that we want to preserve.
+	 */
+	public OutputColsHelper(TableSchema inputSchema, String[] outputColNames, TypeInformation[] outputColTypes,
+							String[] reservedColNames) {
+		this.inputColNames = inputSchema.getFieldNames();
+		this.inputColTypes = inputSchema.getFieldTypes();
+		this.outputColNames = outputColNames;
+		this.outputColTypes = outputColTypes;
+
+		HashSet<String> toReservedCols = new HashSet<>(
+			Arrays.asList(reservedColNames == null ? this.inputColNames : reservedColNames)
+		);
+		//the indices of the columns which need to be reserved.
+		ArrayList<Integer> reservedColIndices = new ArrayList<>(toReservedCols.size());
+		ArrayList<Integer> reservedColToResultIndex = new ArrayList<>(toReservedCols.size());
+		outputColsPosInResult = new int[outputColNames.length];
+		Arrays.fill(outputColsPosInResult, -1);
+		int index = 0;
+		for (int i = 0; i < inputColNames.length; i++) {
+			int key = ArrayUtils.indexOf(outputColNames, inputColNames[i]);
+			if (key >= 0) {
+				outputColsPosInResult[key] = index++;
+				continue;
+			}
+			//add these interested column.
+			if (toReservedCols.contains(inputColNames[i])) {
+				reservedColIndices.add(i);
+				reservedColToResultIndex.add(index++);
+			}
+		}
+		for (int i = 0; i < outputColsPosInResult.length; i++) {
+			if (outputColsPosInResult[i] == -1) {
+				outputColsPosInResult[i] = index++;
+			}
+		}
+		//write reversed column information in array.
+		this.reservedCols = new int[reservedColIndices.size()];
+		this.reservedColsPosInResult = new int[reservedColIndices.size()];
+		for (int i = 0; i < this.reservedCols.length; i++) {
+			this.reservedCols[i] = reservedColIndices.get(i);
+			this.reservedColsPosInResult[i] = reservedColToResultIndex.get(i);
+		}
+	}
+
+	/**
+	 * Get the reserved columns' names.
+	 *
+	 * @return the reserved colNames.
+	 */
+	public String[] getReservedColumns() {
```

Comment:
API is not tested. I made some changes to the test in https://github.com/walterddr/flink/commit/802fa4c6530d90b6322c7be548b0953f17cbf003

Reply:
Thanks.